exchange/exchange.go

```diff
@@ -261,28 +273,26 @@ func (e *exchange) HoldAuction(ctx context.Context, r AuctionRequest, debugLog *
 			}
 		}
 		r.FirstPartyData = resolvedFPD
-		if len(resolvedFPD) > 0 {
-			// rebuild/resync the request in the request wrapper.
-			// it needs to be here to update ReqWrapper and requestExt because req.ext was modified after FPD extraction
-			if err := r.BidRequestWrapper.RebuildRequest(); err != nil {
-				return nil, err
-			}
-			requestExt, err = extractBidRequestExt(r.BidRequestWrapper.BidRequest)
-			if err != nil {
-				return nil, err
-			}
-		}
 	}
 
-	bidAdjustmentFactors := getExtBidAdjustmentFactors(requestExt)
+	bidAdjustmentFactors := getExtBidAdjustmentFactors(requestExtPrebid)
 
-	recordImpMetrics(r.BidRequestWrapper.BidRequest, e.me)
+	recordImpMetrics(r.BidRequestWrapper, e.me)
 
 	// Make our best guess if GDPR applies
-	gdprDefaultValue := e.parseGDPRDefaultValue(r.BidRequestWrapper.BidRequest)
+	gdprDefaultValue := e.parseGDPRDefaultValue(r.BidRequestWrapper)
+
+	// rebuild/resync the request in the request wrapper.
+	if err := r.BidRequestWrapper.RebuildRequest(); err != nil {
```

Comment:
A non-dirty `RequestWrapper` will rapidly exit `request.RebuildRequest()`, but still do at least two levels of stack allocations because of function calls. Because no `BidRequest` fields seem to be modified between line 252 and here, we could slightly optimize by doing a `responseDebugAllow` check instead of an extra `request.RebuildRequest()` call.
```
283     gdprDefaultValue := e.parseGDPRDefaultValue(r.BidRequestWrapper)
284
    +   if !responseDebugAllow {
285         // rebuild/resync the request in the request wrapper.
286         if err := r.BidRequestWrapper.RebuildRequest(); err != nil {
287             return nil, err
288         }
    +   }
296
297     e.me.RecordRequestPrivacy(privacyLabels)
exchange/exchange.go
```
Unless I'm missing something, these are the outcomes I could identify:

**As written**
| ExecuteProcessedAuctionStage calls `RebuildRequest()` | `responseDebugAllow` | times `responseDebugAllow()` gets called |
| ---------------------------- | ------------------ | ---------------------- |
|        no                    |     false          | 1 time. Line 286 |
|        no                    |     true           | 2 times. Lines 252 and 286 |
|       yes                    |     false          | 2 times. Lines 217 and 286 |
|       yes                    |     true           | 3 times. Lines 217, 252 and 286 |

**Adding the `!responseDebugAllow` check**
| ExecuteProcessedAuctionStage calls `RebuildRequest()` | `responseDebugAllow` | times `responseDebugAllow()` gets called |
| ---------------------------- | ------------------ | ---------------------- |
|        no                    |     false          | 1 time on line 286  |
|        no                    |     true           | 1 time. Line 252 |
|       yes                    |     false          | 2 times. Lines 217 and 286 |
|       yes                    |     true           | 2 times. Lines 217 and 252 |


On the other hand, I admit the second `responseDebugAllow` does look a little verbose. What do you think?

Reply:
Thank you for the review of this potential optimization. I appreciate the level of detail you captured. 

As you noted, this optimization relies on there being no modification to the request wrapper after line 252. What if that changes in the future? Do you perhaps have an idea for how to codify this assumption somehow? .. such as with a unit test? Or is this really not a practical concern?

I don't really know why the request wrapper is rebuilt at that specific spot. I wonder if that may be moved entirely.. but now I think I'm getting too complicated. :)

Reply:
Agree. I guess the `ReqestWrapper` is a work in progress and in phases to follow we can decide where should we trigger the rebuilds.

Reply:
> I guess the ReqestWrapper is a work in progress

Yes. Absolutely. @hhhjort is working on a follow-up PR to continue that work.